### DISCUSSION TITLE
How to get code coverage for scripts called as worker process/thread?

### DISCUSSION THREAD

**Discussion by nujarum** [vitest-dev/vitest#3851] [Q&A] [answered]
I have run vitest on functions that call worker process (or worker thread) internally as shown below, but I can't seem to get code coverage for the worker script.

Is there any way (or options) to get coverage of scripts called by [`child_process`](https://nodejs.org/dist/latest/docs/api/child_process.html) or [`worker_threads`](https://nodejs.org/dist/latest/docs/api/worker_threads.html)?

<p><details open>
<summary><code>main-process.ts</code></summary>

```ts
import { fork } from 'node:child_process';
import { once } from 'node:events';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const workerPath = require.resolve('#worker-process');

export async function main() {
    const worker = fork(workerPath);
    try {
        await new Promise<void>((done, fail) => {
            worker.send(Math.random(), err => err ? fail(err) : done());
        });
        const [response] = await once(worker, 'message');
        return response as number;
    } finally {
        worker.removeAllListeners();
        worker.disconnect();
        // worker.kill();
    }
}
```

</details></p>

<p><details open>
<summary><code>worker-process.ts</code></summary>

```ts
process.on('message', async value => {
    await new Promise<void>((done, fail) => {
        process?.send?.(value, (err: unknown) => err ? fail(err) : done());
    });
    process.exit();
});
```

</details></p>

<p><details open>
<summary><code>vite.config.ts</code></summary>

```ts
import { defineConfig } from 'vitest/config';

const esbuildOptions = Object.freeze({
    format: 'esm',
    // minify: true,
    sourcemap: true,
    target: 'node18',
    treeShaking: true,
});

export default defineConfig({
    esbuild: { ...esbuildOptions },
    resolve: {
        alias: [
            { find: /^#(.+)$/, replacement: './src/$1' },
        ],
    },
    test: {
        coverage: {
            provider: 'istanbul',
            // provider: 'v8',
        },
    },
});
```

</details></p>

<p><details open>
<summary>run <code>vitest</code></summary>

```console
$ npx vitest run --coverage ./test/multi-process.spec.ts

 RUN  v0.33.0 /path/to/my/project
      Coverage enabled with istanbul

 ✓ test/multi-process.spec.ts (1)
   ✓ index

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  00:10:15
   Duration  848ms (transform 247ms, setup 0ms, collect 244ms, tests 80ms, environment 0ms, prepare 184ms)

 % Coverage report from istanbul
-----------------|---------|----------|---------|---------|-------------------
File             | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-----------------|---------|----------|---------|---------|-------------------
All files        |     100 |       50 |     100 |     100 | 
 main-process.ts |     100 |       50 |     100 |     100 | 12
-----------------|---------|----------|---------|---------|-------------------
```
↑ Coverage of `worker-process.ts` has not been obtained.

</details></p>

**Comment by nujarum** [comment]
I found a [similar issue registered in Jest](https://github.com/jestjs/jest/issues/5274), but it seems that it has not been resolved yet.

Is this possible in Vitest unlike Jest? Or is there any plan to support it in the future?

**Comment by AriPerkkio** [comment, marked as answer]
Vitest can collect coverage of files that are loaded through Vite. Files that are run using `node:worker_threads`, `node:child_process` or `require()` are not intercepted by Vite. This also means that Vitest mocks or other similar features are not available inside such files.

Even if Vitest used tools like `spawn-wrap` to [overwrite `node:child_process` APIs](https://github.com/istanbuljs/spawn-wrap/blob/22c26aa8d5df11588163ab0de95ece76e67b7140/index.js#L47-L48) just for code instrumentation, we would still need to figure out where to get source maps as Vite would not have those now. I'm not saying it's impossible, but I think it would require quite complex setup. 

There's some related discussion here: https://github.com/vitest-dev/vitest/discussions/2911.

**Reply by nujarum** [reply]
@AriPerkkio Thank you for the detailed explanation. I understand now.

I will try to deal with it by making the scope of the unit test more detailed, etc., instead of doing integration tests.

**Reply by AriPerkkio** [reply]
This will be fixed by
- https://github.com/vitest-dev/vitest/pull/9976

There is a preview build available in https://github.com/vitest-dev/vitest/pull/9976#issuecomment-4135015319.

Sorry it took almost 3 years 🙃
